### PR TITLE
extract common logging functionality into ApiSymbolExtensions

### PR DESCRIPTION
**Pull request by smasher164**
This change refactors APICompat and GenAPI to use a common set of logging abstractions. These include `ILog`, `ISuppressableLog`, `MSBuildLog`, `ConsoleLog`, `SuppressableConsoleLog`, `SuppressableMSBuildLog`, and `SuppressableTestLog`.

```
ILog
├─ConsoleLog────────────┬─SuppressableConsoleLog
├─ISuppressableLog──────┤
│ └─SuppressableTestLog │
└─MSBuildLog────────────┴─SuppressableMSBuildLog
```

Because we don't want the console logger to depend on `Microsoft.Build.*` packages, we exclude files that depend on it from `ApiSymbolExtensions`'s build. The depending packages (in this case `GenAPI.Task` and `ApiCompat.Task`) are required to include those files explicitly.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
The console logger must function without a reference to the msbuild packages. We must not mix frontend stacks. That's why we had the copy of the MessageImportance class to work on both msbuild and the CLI.

That also means that the ApiSymbolExtensions must not have a PackageReference to msbuild packages. We would either need to introduce an additional project ie ApiSymbolExtensions.MSBuild.csproj or just have the msbuild loggers in ApiSymbolExtensions, not compile them and then manually include them in the ApiCompat and GenAPI frontends.

File: src/Microsoft.DotNet.ApiSymbolExtensions/Logging/ILog.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Framework;
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.ApiSymbolExtensions/Logging/ILog.cs:4]: I agree with Viktor, mixing frontends is not the best option. I believe, we must remove the usage of `Microsoft.Build.Framework` from here.

File: src/GenAPI/Microsoft.DotNet.GenAPI/Microsoft.DotNet.GenAPI.csproj [hunk 2]
```diff
@@ -9,10 +9,10 @@
     <_MicrosoftCodeAnalysisVersion>4.0.1</_MicrosoftCodeAnalysisVersion>
     <_MicrosoftCodeAnalysisVersion Condition="'$(DotNetBuildFromSource)' == 'true'">$(MicrosoftCodeAnalysisPackageVersion)</_MicrosoftCodeAnalysisVersion>
   </PropertyGroup>
-
+
```
Comment (andriipatsula) [inline src/GenAPI/Microsoft.DotNet.GenAPI/Microsoft.DotNet.GenAPI.csproj:12]:
```suggestion

```

File: src/GenAPI/Microsoft.DotNet.GenAPI.Task/Microsoft.DotNet.GenAPI.Task.csproj [hunk 3]
```diff
@@ -22,11 +22,17 @@
...
+  <ItemGroup>
+    <Compile Include="..\..\Microsoft.DotNet.ApiSymbolExtensions\Logging\MSBuildLog.cs" />
+  </ItemGroup>
 
   <ItemGroup>
     <!-- ExcludeAssets="Runtime" is being set to avoid adding package references and dependencies of package and project references into the package. -->
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildPackageVersion)" ExcludeAssets="Runtime" />
     <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" ExcludeAssets="Runtime" />
     <ProjectReference Include="..\Microsoft.DotNet.GenAPI\Microsoft.DotNet.GenAPI.csproj" ExcludeAssets="Runtime" />
+    <ProjectReference Include="..\..\Microsoft.DotNet.ApiSymbolExtensions\Microsoft.DotNet.ApiSymbolExtensions.csproj" ExcludeAssets="runtime" />
```
Comment (andriipatsula) [inline src/GenAPI/Microsoft.DotNet.GenAPI.Task/Microsoft.DotNet.GenAPI.Task.csproj:35]: We have a ProjectReference to `Microsoft.DotNet.ApiSymbolExtensions.csproj` in `Microsoft.DotNet.GenAPI`. I think, we can omit this reference here

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/Microsoft.DotNet.ApiCompat.Task.csproj [hunk 4]
```diff
@@ -23,10 +23,16 @@
...
+  <!-- Include MSBuild logger -->
+  <ItemGroup>
+    <Compile Include="..\..\Microsoft.DotNet.ApiSymbolExtensions\Logging\MSBuildLog.cs" />
+  </ItemGroup>
 
   <ItemGroup>
     <!-- ExcludeAssets="Runtime" is being set to avoid adding package references and dependencies of package and project references into the package. -->
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildPackageVersion)" ExcludeAssets="Runtime" />
     <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" ExcludeAssets="Runtime" />
+    <ProjectReference Include="..\..\Microsoft.DotNet.ApiSymbolExtensions\Microsoft.DotNet.ApiSymbolExtensions.csproj" ExcludeAssets="Runtime" />
```
Comment (andriipatsula) [inline src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/Microsoft.DotNet.ApiCompat.Task.csproj:35]: we have reference to `Microsoft.DotNet.ApiSymbolExtensions.csproj` in `Microsoft.DotNet.ApiCompatibility.csproj`. I think we can omit this line here

**Review by andriipatsula — APPROVED** [review]

**Comment by smasher164** [comment]
A lot of the tests are failing with the error `Could not load file or assembly 'Microsoft.CodeAnalysis`. I need to dig in to see if there's a code change here that affected that dependency from being loaded.

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/SuppressableConsoleLog.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Microsoft.DotNet.ApiCompat.Tool
+{
+    /// <summary>
+    /// Class that can log Suppressions to the Console, by implementing ConsoleLog and ISuppressableLog.
+    /// </summary>
+    internal class SuppressableConsoleLog : ConsoleLog, ISuppressableLog
+    {
+        private readonly ISuppressionEngine _suppressionEngine;
+        public bool SuppressionWasLogged { get; private set; }
```
Comment (ViktorHofer) [inline src/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/SuppressableConsoleLog.cs:17]: `/// <inheritdoc />` is missing here

Also please add an empty line to separate the private field from the property (code convention).

File: src/Microsoft.DotNet.ApiSymbolExtensions/Logging/MessageLevel.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.ApiSymbolExtensions.Logging
+{
+    public enum MessageLevel
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/Logging/MessageLevel.cs:6]: Is that a new type? And if yes, what is it needed for?

**Review by ViktorHofer — COMMENTED** [review]
Aside from a few questions, looks great. Thanks

**Pull request merged**